lib/driver.js

```diff
@@ -771,12 +769,63 @@ class XCUITestDriver extends BaseDriver {
     }
     log.debug(`Setting initial orientation to '${orientation}'`);
     try {
-      await this.proxyCommand('/orientation', 'POST', {orientation});
+      await this._executeCommandWithTimeout(B.promisify(this.proxyCommand), 'setScreenOrientation', '/orientation', 'POST', {orientation});
       this.opts.curOrientation = orientation;
     } catch (err) {
       log.warn(`Setting initial orientation failed with: ${err}`);
     }
   }
+
+  _normalizeCommandTimeouts (value) {
+    // The value is normalized already
+    if (typeof value !== 'string') return value;
+
+    let result = {};
+    // Use as default timeout for all commands if a single integer value is provided
+    if (!isNaN(value)) {
+      result[DEFAULT_TIMEOUT_KEY] = _.toInteger(value);
+      return result;
+    }
+
+    // JSON object has been provided. Let's parse it
+    try {
+      result = JSON.parse(value);
+      if (!_.isPlainObject(result)) {
+        throw new Error();
+      }
+    } catch (err) {
+      log.errorAndThrow(`"commandTimeouts" capability should be a valid JSON object. "${value}" was given instead`);
+    }
+    _.forIn(result, (timeout, key) => {
+      if (!_.isInteger(timeout) || timeout <= 0) {
+        log.errorAndThrow(`The timeout for "${key}" should be a valid natural number of milliseconds. "${timeout}" was given instead`);
+      }
+    });
+    return result;
+  }
+
+  async _executeCommandWithTimeout(commandPromise, cmdName, ...args) {
+    try {
+      this.opts.commandTimeouts = this._normalizeCommandTimeouts(this.opts.commandTimeouts);
+    } catch (err) {
+      await this.startUnexpectedShutdown(new errors.UnknownError(err));
+      throw err;
+    }
+    if (_.isPlainObject(this.opts.commandTimeouts)) {
+      const timeout = this.opts.commandTimeouts[cmdName] || this.opts.commandTimeouts[DEFAULT_TIMEOUT_KEY];
+      const errMsg = `Appium did not get any response from "${cmdName}" command in ${timeout} ms`;
+      if (_.isInteger(timeout)) {
+        log.debug(`Setting custom timeout to ${timeout} ms for "${cmdName}" command`);
+        return await commandPromise(...args)
+                 .timeout(timeout)
+                 .catch(B.Promise.TimeoutError, async () => {
+                   await this.startUnexpectedShutdown(new errors.TimeoutError(errMsg));
```

Comment:
The initial was that one will set timeouts for these particular commands, that are subject for potential freeze. So killing the session is OK in that case, since that was done deliberately